### ISSUE TITLE
Should Change "Check by" Dropdown

### ISSUE THREAD

**Issue by excitive** [validator/validator#134]
The "Check by" select drop-down should be changed to either emulated tabs (like in earlier version) or  radio-buttons. A drop-down is not necessary for a choice between only three options. Radios or Tabs would reduce one necessary click as well.

![screen shot 2015-07-26 at 4 59 36 pm](https://cloud.githubusercontent.com/assets/13506996/8893667/c4378bd8-33b7-11e5-95ed-03ae02119bd9.png)

Cheers!

**Comment by sideshowbarker** [comment]
Thanks for using the HTML checker and taking time to suggest an improvement.

> Radios or Tabs would reduce one necessary click as well.

I guess radio buttons would, but I don't see how a tabs-based UI would.

Anyway, beyond possibly reducing one click, what actual important problem do you believe would be solved by changing to a tabs-based UI or radio buttons?

Personally I don't like the tabs UI of the legacy validator and I think it's worse UX. Nor IMHO would radio buttons  be an improvement. Both a tabs-based UI and radio buttons would seem to me to just take up additional screen real-estate without really adding anything for users.

That said, I welcome pull requests with proposed patches for an alternative UI to consider.

**Comment by unor** [comment]
I wanted to suggest this, too. It’s the biggest issue I have with using https://validator.w3.org/nu/ over https://validator.w3.org/.

>> Radios or Tabs would reduce one necessary click as well.

> I guess radio buttons would, but I don't see how a tabs-based UI would.

Tabs and radio buttons would work exactly the same, click-wise: both require only one click if you want to use the non-default option. It’s just that one uses `form` elements, and the other one `a`.

See https://validator.w3.org/, which uses tabs: 0 clicks for selecting "Validate by URI" (because it’s the default), 1 click for selecting "File Upload" or "Direct Input".

> Anyway, beyond possibly reducing one click, what actual important problem do you believe would be solved by changing to a tabs-based UI or radio buttons?

Visibility. Users see which options they have without having to open the drop-down.

**Comment by sideshowbarker** [comment]
> Tabs and radio buttons would work exactly the same, click-wise: both require only one click if you want to use the non-default option.

Same with the select menu; it's just a mouse or keyboard action rather than a click.

> See https://validator.w3.org/, which uses tabs: 0 clicks for selecting "Validate by URI" (because it’s the default), 1 click for selecting "File Upload" or "Direct Input".

See the https://validator.w3.org/nu/, which uses a select menu: 0 clicks for selecting "Check by address" (because it's the default), one keyboard or mouse action for selecting "Check by file upload" or "Check by text input".

If you're next going to launch into an explanation of how choosing something from select control is significantly more burdensome than clicking something—or you're tempted to turn this into a discussion about "clicks" vs other types of UI actions—my answer is going to be that I'm really not interested in having a meta-discussion about which particular classes of UI elements are somehow inherently better than others, or whatever.

> > Anyway, beyond possibly reducing one click, what actual important problem do you believe would be solved by changing to a tabs-based UI or radio buttons?
> Visibility. Users see which basic options they have without having to open the drop-down.

Please. You, me, and everybody else who's ever used any sort of GUI interface all understand that when you see a select menu, it provides some additional options for you to explore. The fact that there are other options available there *is* visible from the nature of the control.

Beyond that I am not interested in having a meta-discussion about… what? Difficulties that a standard select-menu control throws in the way of users?

This UI is using standard features of the platform and not violating any documented UI/UX guidelines that I am aware of.

Beyond that, how we have the UI present options is a judgment call. And in the design of the UI, the developers have made a particular judgment call about it, and provided a sound rationale for the choice (that is, "it wastes less space in the UI").

> These are different "modes", not just some settings only interesting sometimes and/or for advanced users.

Clearly, select menus are not just for advanced users.

Users know what select menus are for, and they know how to use them. If they didn't, there are obviously many other applications in the world they're not going to be able to use.  And I mean everywhere, not just on the Web—because this isn't something specific to the set of UI controls the Web provides. A select control is standard part of any UI toolkit of every GUI platform everywhere.

I have very limited time to spend working on this tool. I very much prefer to spend that time solving real problems for real users. I personally don't see a problem here that needs to be solved. But if somebody else wants to spend time coding up a patch for an alternative UI that uses tabs or radio buttons instead of a select menu, then I would review it in good faith myself, and work with others to evaluate and decide whether it should be merged or not.

**Comment by unor** [comment]
> Same with the select menu; it's just a mouse or keyboard action rather than a click.

If you want to select a non-default option in the drop-down (as mouse user), you need *two* clicks.

> Clearly, select menus are not just for advanced users.

This is not what I meant. I wanted to point out the difference between options/settings that are optional, only sometimes of relevance, only relevant for advanced users, etc., where it can be a good choice to use a select list (e.g., like https://validator.w3.org/ uses them for "Character Encoding"), versus the three different "modes" (for a lack of a better word) that both validators offer: they don’t change some setting of the validation, but they enable using the validator in the first place (for those users that want to upload a file or copy-paste the markup).

**Comment by sideshowbarker** [comment]
> If you want to select a non-default option in the drop-down (as mouse user), you need two clicks.

Fair enough. I stand corrected. But that in no way changes how very little I care about this personally.

> they don’t change some setting of the validation, but they enable using the validator in the first place (for those users that want to upload a file or copy-paste the markup).

It is very much by design that the UI is optimized for the most common case, which is: users who just want to enter a URL for a Web document to check.

That is the the most common case by an extremely wide margin; the usage patterns indicate that there are at least 100 times as many users who enter a URL to check a document as there are users who check by file upload, and that there are at least 50 times as many users who enter a URL as users who directly enter text to check.

I do not believe it's good UI/UX design to give equal prominence in a UI to non-default options that are used 50 to 100 times less often than the default. That's just a waste.

And I do not believe it's a hardship for that smaller set of people who aren't using the default to need to expend literally just some additional milliseconds to choose an option from a select menu and make one additional click.

So as I said in earlier comments, I sincerely don't see any real problem here that needs solving, so I'm not going to spend any time on it. So it's basically not a great use  of time to lobby me further on this here.

At this point, if you feel strongly enough about it, I think you could spend your time more productively by writing a patch for the kind of UI you'd like to have, or by convincing somone else to write such a patch.

**Comment by excitive** [comment]
Sorry for responding late. I would like to quickly summarize the three controls as per my understanding. In an input *form*...

A **radio button** is best when user needs to make a choice between a less number of (3-4) apparent choices which are fixed, not changing. Radios also allow the user to quickly see all available choices before making one.

A **dropdown** is preferable when the number of options are large and are generated dynamically and so the number of options might change. From user’s perspective it is when user needs to make one choice out of a list, e.g. country. Dropdown options are often part of actual data submitted to server rather than UI modification (see below). 

**Tabs** are usually good when the number of choices are finite, should be apparent to user and importantly, making one choice usually *changes* something on screen - UI or content. Users visualize tabs as pages or hidden sub-sections of the page which (s)he can discover by selecting appropriate Tab. Hence Tabs suit the most in our case.

However I agree that Tabs take up some additional real estate and are more suitable when placed on top of the page and not in between. Though even if the number of user base for non-default options is smaller (I myself one of them), if an easier option gets agreed upon, it can still be implemented for the sake of usability.

Here is a quick wireframe that utilizes tabs. I have dared to simplify the design, which takes the user's focus right where it should be, and yet provide all the options:

![validator03](https://cloud.githubusercontent.com/assets/13506996/9080255/9ae4e1ee-3b6f-11e5-85f9-2b5592e169a5.jpg)

Following two are earlier versions before they were simplified to above. I would love to hear the feedback.

![validator01](https://cloud.githubusercontent.com/assets/13506996/9079842/ec51aa7a-3b6b-11e5-870d-05e1b5f3bbf2.jpg)
--
![validator02](https://cloud.githubusercontent.com/assets/13506996/9079846/fa06e86a-3b6b-11e5-92fd-4020075256ff.jpg)

Cheers!

**Comment by sideshowbarker** [comment]
Thanks for taking time to mock up the wireframes. I think a well-designed tabs-based or radio-button-based UI would most likely be an improvement. So I’m not at all arguing against any changes being made. I’m just saying that I’m not going to take time myself to write up code for any changes for this. But as I said way back in my first comment in this issue, at https://github.com/validator/validator/issues/134#issuecomment-124975723, I welcome pull requests with proposed patches for an alternative UI to consider.

An alternative UI could be completely implemented just with changes to https://github.com/validator/validator/blob/master/site/nu-script.js and https://github.com/validator/validator/blob/master/site/nu-style.css. So I would be happy to review a PR with patches to that code.

**Comment by sideshowbarker** [comment]
Closing, as we never got any patch for this and at this point it seems unlikely we’ll be getting one.
